#################################################################################################
#   This script is used to edit the markers and genotype probability arrays generated by 
#     Mandy Chen mandy.chen@jax.org and convert to qtl2 format.
#
#
#   Arrays were obtained on cadillac HPC at /projects/chenm/haplotype/R5/res
#
#
#
#   Input:
#     1.) gigaMuga array 
#     2.) markers for gigaMuga array - http://csbio.unc.edu/MUGA/snps.gigamuga.Rdata
#     2.) gigaMuga interpolated on 69k grid
#     3.) markers for 69k grid: Cadillac - /projects/churchill-lab/resource/muga_reference/marker_grid_0.02cM_plus.rds 
#
# 
#   Output:
#     1.) .Rdata containing gigaMuga genotype probabilities, kinship, map, and markers dataframe
#     2.) .Rdata containing gigaMuga genotype probabilities, kinship, map, and markers dataframe for 69k grid
#
#
#
#   Author: Duy Pham
#   E-mail: duy.pham@jax.org
#################################################################################################


### Options and libraries
options(stringsAsFactors = FALSE)
library(qtl2convert)
library(dplyr)
library(qtl2)





### Inputs:
#    1.) gigaMUGA - genotype probabilities: 219 x 8 x 137,220
#    2.) gigaMUGA69k - genotype probabilities on 69k grid: 219 x 8 x 69,0005
#    3.) snps - for gigaMUGA
#    4.) marker69k - for 69k grid
gigaMuga <- readRDS('~/Desktop/Pazdro Cardiac Hypertrophy/Genotypes/UNC/Robert_Pazdro_Large_File__GigaMUGA_probs_8state.rds')
gigaMuga <- probs_qtl2_to_doqtl(probs = gigaMuga)
load('~/Desktop/Pazdro Cardiac Hypertrophy/Genotypes/UNC/snps.gigamuga.Rdata')

gigaMuga69k <- readRDS('~/Desktop/Pazdro Cardiac Hypertrophy/Genotypes/69k/Robert_Pazdro_Large_File__GigaMUGA_probs_8state_69k.rds')
markers69k  <- readRDS('~/Desktop/Pazdro Cardiac Hypertrophy/Genotypes/69k/marker_grid_0.02cM_plus.rds')










### Converting markers to qtl2 list
markers <- snps %>%
            filter(chr %in% paste0('chr', c(1:19,'X'))) %>% 
            filter(marker %in% dimnames(gigaMuga)[[3]]) %>%
            mutate(chr = gsub('chr', '', chr), chr = factor(chr, levels = c(1:19,'X')), mbp = pos / 1e6) %>%
            rename(marker.id = marker, bp = pos, pos = mbp) %>%
            select(marker.id, chr, pos, bp, cM)
map <- map_df_to_list(map = markers, chr_column = 'chr', pos_column = 'pos', marker_column = 'marker.id')


markers69k <- markers69k %>%
                mutate(chr = factor(chr, levels = c(1:19,'X'))) %>%
                rename(marker.id = marker) %>%
                select(marker.id, chr, pos, bp, cM)
map69k <- map_df_to_list(map = markers69k, chr_column = 'chr', pos_column = 'pos', marker_column = 'marker.id')










### Removing characters in mouse id of genoprobs to contain only 'DO-#'. Convert genotype probabilties to qtl2
dimnames(gigaMuga)[[1]] <- gsub('Large_File_|Small_File_|Wave2_|_[A-Z][0-9]', '', dimnames(gigaMuga)[[1]])
dimnames(gigaMuga69k)[[1]] <- gsub('Large_File_|Small_File_|Wave2_|_[A-Z][0-9]', '', dimnames(gigaMuga69k)[[1]])

gigaMuga    <- probs_doqtl_to_qtl2(probs = gigaMuga, map = markers, chr_column = 'chr', pos_column = 'pos', marker_column = 'marker.id')
gigaMuga69k <- probs_doqtl_to_qtl2(probs = gigaMuga69k, map = markers69k, chr_column = 'chr', pos_column = 'pos', marker_column = 'marker.id')








### Calculating Kinship using 'loco' method
K    <- calc_kinship(probs = gigaMuga, type = 'loco', cores = 10)
K69k <- calc_kinship(probs = gigaMuga69k, type = 'loco', cores = 10)











### Save gigaMUGA .Rdata
genoprobs <- gigaMuga
markers <- markers
map <- map
K <- K
save(genoprobs, markers, map, K, file = '~/Desktop/pazdro_gigaMUGA_genoprobs_qtl2.Rdata')







### Save gigaMUGA on 69k grid .Rdata
genoprobs <- gigaMuga69k
markers <- markers69k
map <- map69k
K <- K69k
save(genoprobs, markers, map, K, file = '~/Desktop/pazdro_69k_genoprobs_qtl2.Rdata')
